src/Control/Distributed/Process/Node/RegistryAgent.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RecursiveDo  #-}
+-----------------------------------------------------------------------------
+---- |
+---- Module      :  Control.Distributed.Process.Node.RegistryAgent
+---- Copyright   :  (c) Tweag I/O 2015
+---- License     :  BSD3 (see the file LICENSE)
+----
+---- Maintainer  :  Tim Watson <watson.timothy@gmail.com>
+---- Stability   :  experimental
+---- Portability :  non-portable (requires concurrency)
+----
+---- This module provides a registry monitoring agent, implemented as a
+---- /distributed-process Management Agent/. Every 'node' starts this agent on
+---- startup. The agent will monitor every remote process that was added to the
+---- local registry, so the node removes the process from the registry when it
+---- dies or when a network failure is detected.
+----
+-------------------------------------------------------------------------------
+
+module Control.Distributed.Process.Node.RegistryAgent
+    ( registryMonitorAgent
+    ) where
+
+import Control.Distributed.Process.Management
+import Control.Distributed.Process.Internal.Types
+import Control.Distributed.Process.Internal.Primitives
+import Data.Foldable (forM_)
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+registryMonitorAgentId :: MxAgentId
+registryMonitorAgentId = MxAgentId "service.registry.monitoring"
+
+-- | Registry monitor agent
+--
+-- This agent listens for 'MxRegistered' and 'MxUnRegistered' events and tracks
+-- all remote 'ProcessId's that are stored in the registry.
+--
+-- When a remote process is registered, the agent starts monitoring it until it
+-- is unregistered or the monitor notification arrives.
+--
+-- The agent keeps the amount of labels associated with each registered remote
+-- process. This is necessary so the process is unmonitored only when it has
+-- been unregistered from all of the labels.
+--
+registryMonitorAgent :: Process ProcessId
+registryMonitorAgent = do
+  nid <- getSelfNode
+  -- For each process the map associates the 'MonitorRef' used to monitor it and
+  -- the amount of labels associated with it.
+  mxAgent registryMonitorAgentId (Map.empty :: Map ProcessId (MonitorRef, Int))
```

Comment:
Oops - late at night pasted wrong code snippet. I meant to paste this one:

```Haskell
        case ev of
          MxRegistered pid label
            | processNodeId pid /= nid -> do
              mref <- liftMX $ monitor pid
              mxUpdateLocal (Map.insert label mref)
          MxUnRegistered pid label
            | processNodeId pid /= nid -> do
              mrefs <- mxGetLocal
              forM_ (label `Map.lookup` mrefs) $ \mref -> do
                liftMX $ unmonitorAsync mref
                mxUpdateLocal (Map.delete label)
          _ -> return ()
```
i.e. as the previous comment was saying, one monitor ref per label, not per process. The only difference with the previous snippet is the key of the map. Everything else is the same. BUT, there is a complication, as you point out. It's just that it's not in the `MxUnregistered` case but in the `ProcessMonitorNotification` case, which when written out becomes something like:

```Haskell
mxSink $ \(ProcessMonitorNotification mref _ _) -> do
        mxUpdateLocal $ Map.filter (== mref)
        mxReady
```

I assume you meant to write `ProcessMonitorNotification` instead of `MxUnregistered`. The code is just as simple, but note the linear operation, which I also assume is what you were referring to. It's the same linear operation that's already [here](https://github.com/haskell-distributed/distributed-process/blob/master/src/Control/Distributed/Process/Node.hs#L812), over the exact same set of labels. Processes shouldn't be overusing the registry to register thousands of processes, so that's just about ok.